### PR TITLE
Add creating many objects in one request

### PR DESCRIPTION
**Pull request by mimo00**
This change is [Reviewable](https://reviewable.io/reviews/socialwifi/flask-jsonapi/9)

### CODE REVIEW

File: flask_jsonapi/resource_repositories.py [hunk 1]
```diff
@@ -40,6 +42,10 @@ def __init__(self, *, repository=None, schema=None, filter_schema=None, view_dec
             self.view_decorators = view_decorators
         if view_kwargs:
             self.view_kwargs = view_kwargs
+        if nested:
+            self.nested = nested
+        if self.nested:
+            self.extend_repository()
```
Comment (mimo00) [inline flask_jsonapi/resource_repositories.py]: if self.nested:
+            self.repository = self.extend_repository()

File: flask_jsonapi/resources.py [hunk 2]
```diff
@@ -119,6 +122,7 @@ def destroy(self, id):
 class ResourceList(ResourceBase):
     methods = ['GET', 'POST']
     filter_schema = filters_schema.FilterSchema({})
+    id_map = {}
```
Comment (mimo00) [inline flask_jsonapi/resources.py]: delete

File: flask_jsonapi/resources.py [hunk 3]
```diff
@@ -154,9 +158,13 @@ def post(self, *args, **kwargs):
                 return self.prepare_response(data)
 
     def prepare_response(self, data):
```
Comment (mimo00) [inline flask_jsonapi/resources.py]: change attribute nested to class Inheritance.

File: flask_jsonapi/nested_repository.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+class NestedRepository:
+    children_repositories = {}
+
+    def __init__(self, repository):
+        self.repository = repository
+        try:
+            self.children_repositories = repository.children_repositories
+        except AttributeError:
+            self.children_repositories = {}
```
Comment (mimo00) [inline flask_jsonapi/nested_repository.py]: raise Exception("Lack of children_repositories")

File: flask_jsonapi/nested_repository.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        return all(child_field in data.keys() for child_field in self.children_repositories.keys())
+
+    def get_model_dict(self, dict_with_nested_object):
+        model_dict = dict_with_nested_object.copy()
+        for child_field in self.children_repositories.keys():
+            model_dict.pop(child_field)
+        return model_dict
+
+    def create_children_models(self, model, data,  **kwargs):
+        for child_field in self.children_repositories.keys():
```
Comment (mimo00) [inline flask_jsonapi/nested_repository.py]: for key, value

File: flask_jsonapi/nested_repository.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+            for child_tree in data.get(child_field):
+                parent_foreign_key_name = self.children_repositories[child_field].foreign_parent_name
+                child_tree[parent_foreign_key_name] = model.id
+                __id__ = IdMapper.pop_id(child_tree)
```
Comment (mimo00) [inline flask_jsonapi/nested_repository.py]: __id__ change name

Comment (mimo00) [reply flask_jsonapi/nested_repository.py]:
```
mapper = IdMapper(kwargs['id_map'])
child_tree = mapper.remove_id(child_tree)

class IdMapper
  init

  def remove_id(child_tree):
     child_tree = child_tree.copy()
    self.id =     child_tree.pop(self.temp_id_name)
    return child_tree
```

Comment (mimo00) [reply flask_jsonapi/nested_repository.py]: Think IdMapper= contextmanager

File: flask_jsonapi/marshmallow_nested_extension/schema.py [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+
+class IdMappingSchema:
+    __id__ = fields.Str(required=False)
+
+    def dump(self, obj, many=None, update_fields=True, **kwargs):
+        self.id_map = kwargs.pop('id_map')
+        if self.id_map is not None:
+            if hasattr(obj, 'id') and obj.id in self.id_map.keys():
+                obj.__id__ = self.id_map[obj.id]
+                del self.id_map[obj.id]
```
Comment (mimo00) [inline flask_jsonapi/marshmallow_nested_extension/schema.py]: delete del

File: tests/test_integration.py [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+        self.name = kwargs['name']
+        self.parent_id = kwargs['parent_id']
+
+
+class ParentModel:
+    def __init__(self, **kwargs):
+        self.id = kwargs['id']
+
+
+data_base_simulation = {}
```
Comment (mimo00) [inline tests/test_integration.py]: database

**Comment by jgoclawski** [comment]
Reviewed 4 of 4 files at r1, 1 of 1 files at r2, 4 of 4 files at r3, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[.idea/vcs.xml, line 1 at r1](https://reviewable.io:443/reviews/socialwifi/flask-jsonapi/9#-KxhnwqjbkAO7CVMKj01:-KxhnwqjbkAO7CVMKj02:buyxtvo) ([raw file](https://github.com/socialwifi/flask-jsonapi/blob/ab062a07dc666e5e5c457cc9f8a5dd9056f5e6b3/.idea/vcs.xml#L1)):*
> ```XML
> <?xml version="1.0" encoding="UTF-8"?>
> ```

Don't commit this file.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/socialwifi/flask-jsonapi/9)*

**Comment by jgoclawski** [comment]
:lgtm:

---

Reviewed 8 of 8 files at r5, 1 of 1 files at r6.
Review status: 1 of 9 files reviewed at latest revision, 1 unresolved discussion.

---

*[.idea/vcs.xml, line 1 at r1](https://reviewable.io:443/reviews/socialwifi/flask-jsonapi/9#-KxhnwqjbkAO7CVMKj01:-Kxi-9RtZSSsIL6oY_v3:b32w0x) ([raw file](https://github.com/socialwifi/flask-jsonapi/blob/ab062a07dc666e5e5c457cc9f8a5dd9056f5e6b3/.idea/vcs.xml#L1)):*
<details><summary><i>Previously, jgoclawski (Jakub Gocławski) wrote…</i></summary><blockquote>

Don't commit this file.
</blockquote></details>

OK

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/socialwifi/flask-jsonapi/9#-:-KxhzourJdTyEpqdcAHE:bnfp4nl)*

**Comment by jgoclawski** [comment]
Reviewed 4 of 4 files at r7, 3 of 3 files at r8, 1 of 1 files at r9.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/socialwifi/flask-jsonapi/9)*

**Pull request merged**